File: sdk/python/tests/test_client.py
import hashlib
import pytest

from evidence_api.tdx.quote import TdxQuote

from dstack_sdk import DstackClient, AsyncDstackClient, GetKeyResponse, GetQuoteResponse, GetTlsKeyResponse

def test_sync_client_get_key():
    client = DstackClient()
    result = client.get_key()
    assert isinstance(result, GetKeyResponse)
    assert isinstance(result.decode_key(), bytes)
    assert len(result.decode_key()) == 32

def test_sync_client_get_quote():
    client = DstackClient()
    result = client.get_quote('test')
    assert isinstance(result, GetQuoteResponse)

def test_sync_client_get_tls_key():
    client = DstackClient()
    result = client.get_tls_key()
    assert isinstance(result, GetTlsKeyResponse)
    assert isinstance(result.key, str)
    assert len(result.key) > 0
    assert len(result.certificate_chain) > 0

@pytest.mark.asyncio
async def test_async_client_get_key():
    client = AsyncDstackClient()
    result = await client.get_key()
    assert isinstance(result, GetKeyResponse)

@pytest.mark.asyncio
async def test_async_client_get_quote():
    client = AsyncDstackClient()
    result = await client.get_quote('test')
    assert isinstance(result, GetQuoteResponse)

@pytest.mark.asyncio
async def test_async_client_get_tls_key():
    client = AsyncDstackClient()
    result = await client.get_tls_key()
    assert isinstance(result, GetTlsKeyResponse)
    assert isinstance(result.key, str)
    assert result.key.startswith('-----BEGIN PRIVATE KEY-----')
    assert len(result.certificate_chain) > 0

@pytest.mark.asyncio
async def test_tls_key_uniqueness():
    """Test that TLS keys are unique across multiple calls."""
    client = AsyncDstackClient()
    result1 = await client.get_tls_key()
    result2 = await client.get_tls_key()
    # TLS keys should be unique for each call
    assert result1.key != result2.key

@pytest.mark.asyncio
async def test_replay_rtmr():
    client = AsyncDstackClient()
    result = await client.get_quote('test')
    # TODO evidence_api is a bit out-of-date, we need an up-to-date implementation.
    tdxQuote = TdxQuote(bytes.fromhex(result.quote))
    rtmrs = result.replay_rtmrs()
    assert rtmrs[0] == tdxQuote.body.rtmr0.hex()
    assert rtmrs[1] == tdxQuote.body.rtmr1.hex()
    assert rtmrs[2] == tdxQuote.body.rtmr2.hex()
    assert rtmrs[3] == tdxQuote.body.rtmr3.hex()

@pytest.mark.asyncio
async def test_get_quote_raw_hash_error():
    with pytest.raises(ValueError) as excinfo:
        client = AsyncDstackClient()
        await client.get_quote('0' * 65)
    assert '64 bytes' in str(excinfo.value)
    with pytest.raises(ValueError) as excinfo:
        client = AsyncDstackClient()
        await client.get_quote(b'0' * 129)
    assert '64 bytes' in str(excinfo.value)

@pytest.mark.asyncio
async def test_report_data():
    reportdata = 'test'
    client = AsyncDstackClient()
    result = await client.get_quote(reportdata)
    tdxQuote = TdxQuote(result.decode_quote())
    reportdata = reportdata.encode('utf-8') + b'\x00' * (64 - len(reportdata))
    assert reportdata == tdxQuote.body.reportdata

File: sdk/rust/tests/test_client.rs
use dcap_qvl::quote::Quote;
use dstack_sdk::dstack_client::DstackClient as AsyncDstackClient;

#[tokio::test]
async fn test_async_client_get_key() {
    let client = AsyncDstackClient::new(None);
    let result = client.get_key(None, None).await.unwrap();
    assert!(!result.key.is_empty());
    assert_eq!(result.decode_key().unwrap().len(), 32);
}

#[tokio::test]
async fn test_async_client_get_quote() {
    let client = AsyncDstackClient::new(None);
    let result = client.get_quote("test".into()).await.unwrap();
    assert!(!result.quote.is_empty());
}

#[tokio::test]
async fn test_async_client_get_tls_key() {
    let client = AsyncDstackClient::new(None);
    let key_config = dstack_sdk::dstack_client::TlsKeyConfig::builder().build();
    let result = client.get_tls_key(key_config).await.unwrap();
    assert!(result.key.starts_with("-----BEGIN PRIVATE KEY-----"));
    assert!(!result.certificate_chain.is_empty());
}

#[tokio::test]
async fn test_tls_key_uniqueness() {
    let client = AsyncDstackClient::new(None);
    let key_config_1 = dstack_sdk::dstack_client::TlsKeyConfig::builder().build();
    let key_config_2 = dstack_sdk::dstack_client::TlsKeyConfig::builder().build();
    let result1 = client.get_tls_key(key_config_1).await.unwrap();
    let result2 = client.get_tls_key(key_config_2).await.unwrap();
    assert_ne!(result1.key, result2.key);
}

#[tokio::test]
async fn test_replay_rtmr() {
    let client = AsyncDstackClient::new(None);
    let result = client.get_quote("test".into()).await.unwrap();
    let rtmrs = result.replay_rtmrs().unwrap();
    let quote = result.decode_quote().unwrap();

    let tdx_quote = Quote::parse(&quote).unwrap();
    let quote_report = tdx_quote.report.as_td10().unwrap();
    assert_eq!(rtmrs[&0], hex::encode(quote_report.rt_mr0));
    assert_eq!(rtmrs[&1], hex::encode(quote_report.rt_mr1));
    assert_eq!(rtmrs[&2], hex::encode(quote_report.rt_mr2));
    assert_eq!(rtmrs[&3], hex::encode(quote_report.rt_mr3));
}

#[tokio::test]
async fn test_report_data() {
    let report_data = "test";
    let client = AsyncDstackClient::new(None);
    let result = client.get_quote(report_data.into()).await.unwrap();
    let quote = result.decode_quote().unwrap();

    let tdx_quote = Quote::parse(&quote).unwrap();
    let quote_report = tdx_quote.report.as_td10().unwrap();
    let expected = {
        let mut padded = report_data.as_bytes().to_vec();
        padded.resize(64, 0);
        padded
    };
    assert_eq!(&quote_report.report_data[..], &expected[..]);
}
